### ISSUE TITLE
Rotation code is UB for zero shift counts

### ISSUE THREAD

**Issue by travisdowns** [nemequ/portable-snippets#13]
If I understand it correctly existing portable rotation code will emit a shift count of 32 if you rotate a 4-byte value by 0:

```
#define PSNIP_BUILTIN_ROTL_DEFINE_PORTABLE(f_n, T, ST)	\
  PSNIP_BUILTIN__FUNCTION				\
  T psnip_intrin_##f_n(T value, ST shift) {		\
    return						\
      (value >> ((sizeof(T) * 8) - shift)) |		\
      (value << shift);					\
}
```

Shifts of amounts _equal_ or greater than the size of the underlying value are UB in C and C++, so a shift of 32 is not legal here.

[Here](https://stackoverflow.com/a/776523/149138) is a good summary of portable, legal ways to implement this. They are generally just as fast (i.e., recognized) on current compilers, but using it on older compilers may be slow.

**Comment by nemequ** [comment]
One thing we have to be careful to avoid is widening the contract of the intrinsic we're emulating.  There are a lot of intrinsics for which passing invalid parameters yields undefined behavior, and I'd actually rather have the portable versions exhibit undefined behavior in those situations as well.  That way at least people will see warnings from tools like ubsan.

I don't see anything [on MSDN](https://msdn.microsoft.com/en-us/library/5cc576c4.aspx) or in [Intel's intrinsics guide](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=rotl&expand=4431) about which values are defined…  In this case I guess the canonical source would be the `rol` x86 instruction, I'm sure they have documentation somewhere about the values that instruction supports…

**Comment by travisdowns** [comment]
There is no UB behavior at the assembly level, so in general a shift by any multiple of 32 is equal to 0 for 32 arguments, and similarly for 64 and 64 bit arguments. In any case I'm talking about a _rotate by zero_ - not about a rotate by 32. That is inherently well defined and would never be a problem (intrinsics or even in languages that offer a method to do).

The problem is really that a **rotate by zero** results in **a shift by 32* due to the snippet of C code used. That's a well known "gotcha" UB in attempts to write a portable rotate function. The code in the linked SO answer should be equivalent to the hardware behavior anyway, so it's strictly better than the existing solution in a correctness sense (including for larger shift counts).

**Comment by travisdowns** [comment]
As an aside, I'd argue that portable snippets should be also trying to offer reasonable behavior for common idioms over (or addition to) emulating a specific intrinsic. For one thing, if platforms offer different contracts for their intrinsics, you are stuck emulating all of them, and then the end user often won't know which one to choose (e.g., if the contract of more than one intrinsic is sufficient for his purposes).

At least when this divergence occurs, it is probably reasonable to pick a behavior which is (a) easily implementable across platforms (b) captures most of what a typical user might want, and offer that. Usually that will be one of the existing instrinsics, but I guess sometimes it won't? Anyway, it's probably better to discuss it somewhere else.

**Comment by nemequ** [comment]
> The code in the linked SO answer should be equivalent to the hardware behavior anyway

Okay, if that's the case then I'm fine with it; I'll put together a patch when I get home tonight, unless you want to do it.

> As an aside, I'd argue that portable snippets should be also trying to offer reasonable behavior for common idioms over (or addition to) emulating a specific intrinsic.

That sounds nice in theory, but I don't think it's really feasible for a lot of intrinsics.  Unfortunately it can cause quite a performance hit on some platforms.  Returning `val` if `shift == 0` is already a pretty big hit on platforms where branching is slow.  I'd rather keep the contract pretty narrow and make sure

FWIW, there is a pretty good presentation at https://www.youtube.com/watch?v=yG1OZ69H_-o which goes into detail on a lot of these issues.  It's mostly from the perspective of the C++ language, but still very much the same problem.

One thing I would be interested in (and probably should have done from the start) is adding assertions to validate the input instead of relying on stuff like ubsan.  As long as they can be hidden behind an `NDEBUG` (or something similar) there is no run-time cost…

**Comment by travisdowns** [comment]
>>    The code in the linked SO answer should be equivalent to the hardware behavior anyway

> Okay, if that's the case then I'm fine with it; 

Well the key point is actually noted after that: the problem occurs for a rotate of zero, which is well-defined everywhere that I've seen (intrinsics, hardware, etc). There are actually _other_ UB behavior that occurs with large rotate counts that may actually also result in large shift counts which are also UB, and the linked solution solves those - but here I was focusing on the straightforward case of a zero rotate where it is clear what the documented intrinsic and hardware behavior is.

**Comment by travisdowns** [comment]
> That sounds nice in theory, but I don't think it's really feasible for a lot of intrinsics. Unfortunately it can cause quite a performance hit on some platforms. Returning val if shift == 0 is already a pretty big hit on platforms where branching is slow. I'd rather keep the contract pretty narrow and make sure.

To be fair, the example is a bad one: I don't know of _any_ platform that can't handle `shift == 0`, as I tried to make clear above. The problem arises with code that implements rotate, that has something like `32 - shift` which results in a shift amount of 32 when `shift == 0`.

The point for this example is that here most/all platforms have well defined behavior for _any_ shift count, even though C and C++ don't. Generally you'll find that the underlying operation `(uint32_t)x >> 32` is either zero or `x`, or [something weird on ARM](https://david.wragg.org/blog/2012/11/shift-instructions.html). So you take a reasonable guess, considering platform support and actual use, that the `& 0xFFFF` behavior (i.e., masking the shift count down to 32 bits), is a reasonable one and implement that as a `psnip` function.

The behavior of the C/C++ committees is basically that any operation that could conceivably be implemented in different ways should have _undefined behavior_ (and, rarely, _implementation defined_ behavior). In cases, for example, where there are only two reasonable choices, the committee still usually chooses to invoke UB. Decisions made in the past are generally carried forward indefinitely into the future as well, so current behavior is constrained by very usual platforms that essentially don't exist any more (and even then, portable-snippets could handle them specifically).

Portable-snippets, on the other hand, can do much better. It can make the 99% case the contact (e.g., arithmetic right shifts), and then simply guard that with checks that you are on a p

This works because C and C++ were used across a huge variety of implementations, including many that are extinct now, but were important (enough) when the standards were written. The general behavior of the committees is to "UB" any behavior that has more than one reasonable implementation. So now you are left with UB behavior for things like signed overflow which are generally well defined on all current interesting platforms. You are also left with UB or IB behavior for things like signed right shifts which are also well defined on all interesting platforms. So the magic of portable-snippets is that it can simply whitelist the interesting platforms (e.g., x86 windows/unix-like, ARM windows, ARM android, POWER, a few more minor ones) with their actual behavior, and then use a C/C++ compliant source version for the rest, which will be completely portable in theory, and very fast in practice.


> 
> FWIW, there is a pretty good presentation at https://www.youtube.com/watch?v=yG1OZ69H_-o which goes into detail on a lot of these issues. It's mostly from the perspective of the C++ language, but still very much the same problem.
> 
> One thing I would be interested in (and probably should have done from the start) is adding assertions to validate the input instead of relying on stuff like ubsan. As long as they can be hidden behind an NDEBUG (or something similar) there is no run-time cost…

**Comment by nemequ** [comment]
> Portable-snippets, on the other hand, can do much better. It can make the 99% case the contact (e.g., arithmetic right shifts), and then simply guard that with checks that you are on a p

It's a good idea, but not *quite* the goal of the builtin module.  The builtin module is about portable implementations of existing built-ins/intrinsics.

In retrospect it *may* have been better (definitely simpler) to implement what you're suggesting, but there are some advantages to simply creating portable versions of existing intrinsics.  For one thing, it's trivial to port existing code; just include a header file and either add a `psnip_intrin/builtin_` prefix or just  define `PSNIP_BUILTIN_EMULATE_NATIVE`.  It's also nice that I didn't need to reinvent the whole API, and could instead defer to people with (theoretically) more experience with different hardware and more familiarity with the use cases.

If you're interested, I'd be willing to add some additional functions (or maybe a new module?) which do what you're talking about.  The `__builtin_*` functions (i.e., the GNU intrinsics) are already designed with portability in mind, so thinking about it as a supplement to that API would probably be a good idea.  That could also provide a good proving ground for potential new GCC built-ins, especially if we can add fast implementations for other archs…

Anyways, it seems like `_rotl`/`_rotr` shouldn't have undefined behavior on x86, so psnip's version shouldn't either.  The staging branch has a version which should work for all inputs… I'm pretty comfortable with the fix for this bug, just want to test the stuff for the other bug you filed a bit more (mostly on ARM) before I push to master.

> Even to use ubsan, you have to compile the all the various versions of the code to test…

Yeah, that's a good point.  I really should have added assertions to make testing easier.  I'll try to do that over the next day or two.

**Comment by nemequ** [comment]
Fixed with 06be7bc91163bb8cf262ff2e74a9c96a5f7a3ad6.

**Comment by travisdowns** [comment]
> It's a good idea, but not quite the goal of the builtin module. The builtin module is about portable implementations of existing built-ins/intrinsics.

Fair enough. 

I can definitely see both sides. I personally just want a library that implements all the various "not in C but there is a good way to do them which varies by platform" in the best, fastest and simplest way. I usually use `gcc` so "implementing the `gcc` intrinsics" (one of the goals of this library) comes pretty close, I guess, but usually my thought is "find a way to find the first set bit" not "emulate the `__builtin_ffs` instruction". In particular, many builtins have existed since forever, and have definitions which made sense for the lowest-common denominator hardware of the day, but today may be unnecessarily loose or strict.

Now I promise I kind of picked `__builtin_ffs` off the top of my head, but it's a good example. It is defined as:

> Returns one plus the index of the least significant 1-bit of x, or if x is zero, returns zero. 

The whole _plus one_ thing is to work around the the fact that there isn't an obvious value to return if there are _no_ bits set (what's the index of the "first bit" in that case?) so they chose `0` for that case, which means the rest of the return values have to be shifted over by one. 

As far as I know, real hardware in common use today never implements this semantic. The "old" x86 instructions `bsf` and `bsf` just ignore the no-bits-set case (undefined result) and return the index, while most of the world has settled on "count trailing zeros" (and "count leading zeros") which returns 32 for a 32-bit value if nothing is set. Intel introduced exactly such `tzcnt` and `lzcnt` instructions, and both Power and ARM have at least "count leading zeros" from which you can derive `ctz`. 

So it is entirely reasonable to provide `ctz` and `clz` functions, since these map naturally to hardware. Even with the "impedance mismatch" of needing to implement it in terms of `__builtin_ffs` it generally works after the optimizer gets a hold of it (I have looked into this specifically). 

Note that `gcc` _also_ offers a `__builtin_ctz` and `__builtin_clz`, but they _also_ have the undefined behavior at 0:

> Returns the number of trailing 0-bits in x, starting at the least significant bit position. If x is 0, the result is undefined. 

That's really weird because the CTZ and CLZ operations are unambiguously well-defined at 0, including on real hardware. This is a "legacy" of the x86 `bsf` and `bsr` instructions which were used to implement these instructions on the very important x86 platform (and which late introduced better instructions). 

> could instead defer to people with (theoretically) more experience with different hardware and more familiarity with the use cases.

First, don't sell yourself short; also, even if you don't think you know the use cases, users can help you. 

Beyond that however, anyone implementing portable-snippets in 2017 has the benefit of a huge amount of hindsight, and is able to understand what works today. Many decisions about intrinsics were made a dozen years ago or more, and often based on one very narrow use-case (the person who filed a performance bug, for example) and then the compilers are stuck supporting that method. With the benefit of time, we can see if those decisions were good ones (often they are) or if there were shortfalls. 

Now I totally understand that a big advantage of the existing approach is that someone who has already written software targeting platform X with its native intrinsics, can now easily port to Y and Z by "dropping in" p-s, either without changing the source, or with a basically rote text-replace on the intrinsic names.

I'm just biased because I'm in the group that just wants to write portable software on X, Y and Z to begin with (and even the "port" people may become these people after) in which case the focus on platform specific intrinsics is a bit odd: I don't really want to look at the lists of intrinsics for MSVC, gcc, clang, etc - I want to look at the list of _functions_ provided by portable-snippets and use that. The goal of supporting all the intrinsics also implies a lot of duplication: many of the intrinsics will be offered by _all_ compilers, sometimes with the same name, but often with different name (and perhaps sometimes with the same name but different behavior - uh oh) - so users will get the pick somewhat arbitrarily between say the "MSVC" and "gcc & friends" naming conventions. I might stick with gcc, but then for the 10% of intrinsics that only MSVC offers, I have to UseANameLikeThis because MSVC does.

Either way, I'm happy with the functionality p-s adds, I just hope not to see it tied _too_ closely to the existing variable quality intrinsics with their locked-in-stone legacy semantics. The problems are still rarer than the ones that "just work", however.
